contractcourt/channel_arbitrator.go

```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
 		if err != nil {
 			return err
 		}
+
+		return nil
+	}
+
+	// Calculate the deadline. This is only valid for sweeping the local
+	// anchor.
+	deadline, err := c.findCommitmentDeadline(heightHint)
+	if err != nil {
+		return err
+	}
+
+	for _, anchor := range anchors {
+		// The first item is the anchor of the local commitment tx, we
+		// will use the deadline calcuated from the active HTLCs.
+		if err := sweepWithDeadline(anchor, deadline); err != nil {
+			return err
+		}
+
+		// For the rest anchors, use the default conf target.
+		deadline = uint32(anchorSweepConfTarget)
 	}
 
 	return nil
 }
 
+// findCommitmentDeadline finds the deadline for a commitment transaction by
+// extracting the minimum CLTV from its HTLCs. Currently only the local
+// commitment transaction is suitable to have a confirmation deadline. And it's
+// defined to be the smaller of,
+//   - the least CLTV from outgoing HTLCs, or,
+//   - the least CLTV from incoming HTLCs if the preimage is available.
+func (c *ChannelArbitrator) findCommitmentDeadline(
+	heightHint uint32) (uint32, error) {
+
+	minHeight := uint32(math.MaxUint32)
+
+	// We only use the HTLC set from our local commitment transaction.
+	htlcs := c.activeHTLCs[LocalHtlcSet]
+
+	// First, iterate through the outgoing HTLCs to find the lowest CLTV
+	// value.
+	for _, htlc := range htlcs.outgoingHTLCs {
+		// Skip if the HTLC is dust.
+		if htlc.OutputIndex < 0 {
+			log.Debugf("ChannelArbitrator(%v): skipped deadline "+
+				"for outgoing dust htlc=%x",
+				c.cfg.ChanPoint, htlc.RHash[:])
+
+			continue
+		}
+
+		if htlc.RefundTimeout < minHeight {
+			minHeight = htlc.RefundTimeout
+		}
+	}
+
+	// Then going through the incoming HTLCs, and update the minHeight when
+	// conditions met.
+	for _, htlc := range htlcs.incomingHTLCs {
+		// Skip if the HTLC is dust.
+		if htlc.OutputIndex < 0 {
+			log.Debugf("ChannelArbitrator(%v): skipped deadline "+
+				"for outgoing dust htlc=%x",
+				c.cfg.ChanPoint, htlc.RHash[:])
+
+			continue
+		}
+
+		preimageAvailable, err := c.isPreimageAvailable(htlc.RHash)
+		if err != nil {
+			return 0, err
+		}
+
+		if !preimageAvailable {
+			continue
+		}
+
+		if htlc.RefundTimeout < minHeight {
+			minHeight = htlc.RefundTimeout
+		}
+	}
+
+	// Calculate the deadline. Notice that our minHeight may never get
+	// updated, which means the local active HTLCs set is empty. We will
+	// then fall back to the default value.
+	deadline := minHeight - heightHint
```

Comment:
FWIW, we could make the deadline a bit more forgiving my also incorporating the CLTV delta for a given outgoing HTLC, as even once the HTLC expires, we still have that delta period before it enters race condition territory. However I like having it a bit stricter to start w/, in the end for non time contested HTLCs, users should see _much_ lower chain fees as a result of force close attempts.

Reply:
Yeah I think this is fine for now, we can always introduce _more_ slack later.